### ISSUE TITLE
external const factory constructors aren't allowed but should be

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#29502]
If I'm reading the grammar right, this should be a syntax error:

```dart
  external const factory bool.fromEnvironment(String name,
      {bool defaultValue: false});
```

Factory constructor syntax is specified by:

```
factoryConstructorSignature:
  factory identifier (‘.’ identifier)? formalParameterList
  ;
```

That rule is only used by:

```
methodSignature:
  constructorSignature initializers? |
  factoryConstructorSignature |
  static? functionSignature |
  static? getterSignature |
  static? setterSignature |
  operatorSignature
;
```

There's no mention of `external` there. All of the productions that use `external` are in:

```
declaration:
  constantConstructorSignature (redirection | initializers)? |
  constructorSignature (redirection | initializers)? |
  external constantConstructorSignature |
  external constructorSignature |
  ((external static ?))? getterSignature |
  ((external static?))? setterSignature |
  external? operatorSignature |
  ((external static?))? functionSignature |
  static (final | const) type? staticFinalDeclarationList |
  final type? initializedIdentifierList |
  static? (var | type) initializedIdentifierList
  ;
```

Probably just need to add `external factoryConstructorSignature` as a clause to `declaration`.

**Comment by eernstg** [comment]
Amazingly, the grammar in the language specification doesn't allow for const factory constructors at all (no matter whether it's external). The problem had not been fixed in Dart.g in the [ANTLR grammar CL](https://codereview.chromium.org/2688903004). I changed `factoryConstructorSignature` in Dart.g to include an optional `const`, which enables const factory constructors in general, and then I added a new case to `declaration` which uses this to enable the declaration of external factory constructors (const as well as non-const).

**Comment by eernstg** [comment]
No milestone now: This correction is part of a larger update of grammar related parts of the specification, based on [Dart.g](https://github.com/dart-lang/sdk/blob/master/docs/language/Dart.g), where the correction was made in June 2017 as mentioned above.

**Comment by eernstg** [comment]
Closing: Fixed in e163637d1537ccc9ccde1d25ca87b82286aa634d.
